### PR TITLE
[WIP] Add support for server-side stacks

### PR DESCRIPTION
**Pull request by dhiltgen**
**This PR is currently marked WIP, because work in the underlying docker/stacks repository is still ongoing.**

Directly related to https://github.com/moby/moby/pull/38832

Contacts for this PR are myself, @dperny and @alexmavr 

**- What I did**

Add support for server-side stacks by importing and wiring up the https://github.com/docker/stacks libraries.

The new server-side stacks support aims to move as much business logic dealing with stacks to the server side as possible in order to achieve a consistent experience across all clients (CLI, GUI, etc.) while bringing the benefits of a 1st-class knowledge of stacks on the server with a reconciler to ensure the actual state matches the desired state for the stack.

**- How I did it**

* Update vendoring to include the new stack types, and client API definitions for server side handling of stacks operations
* Move existing client-side logic into "legacy" subdirs for eventual deprecation and removal.
* Toggle client-side vs. server-side logic based on API version
* New client-side logic leveraging the server-side implementations for stack operations

**- How to verify it**

* Combine with https://github.com/moby/moby/pull/38832 for end-to-end usage
* Unit test coverage is included in this PR
* end-to-end tests are being developed in the https://github.com/docker/stacks repo for inclusion in the broader end-to-end Docker platform tests

**- Description for the changelog**

Add support for server-side stacks

**- A picture of a cute animal (not mandatory but encouraged)**

This is my dog Trinity, playing along with her human siblings
mvimg_20171230_140416_1_exported_1625781761128929608 1

### CODE REVIEW

File: vendor.conf [hunk 1]
```diff
@@ -101,3 +102,35 @@ k8s.io/client-go kubernetes-1.11.2
 k8s.io/kube-openapi d8ea2fe547a448256204cfc68dfee7b26c720acb
 k8s.io/kubernetes v1.11.2
 vbom.ml/util 256737ac55c46798123f754ab7d2c784e2c71783
+
+#get libnetwork packages
+github.com/docker/libnetwork 1a06131fb8a047d919f7deaf02a4c414d7884b83
```
Comment (thaJeztah) [inline vendor.conf]: curious why all these dependencies are now needed on the client side 🤔

Comment (dhiltgen) [reply vendor.conf]: Yeah, I was bummed this pulled in so much but it comes from the code just below this comment block:  https://github.com/docker/cli/pull/1721/files#diff-04999fb6f01d1478d54e2713a0b41a15R113

Comment (thaJeztah) [reply vendor.conf]: I also noticed here that we're pulling in the backend; https://github.com/docker/cli/blob/99c7e4210fa49f65f82cacf7e5f728d7a2e247c0/vendor/github.com/docker/stacks/pkg/interfaces/interfaces.go#L6-L37

Many of these should not be in the CLI (and might not even work on macOS, Windows)

Comment (vdemeester) [reply vendor.conf]: If those dependencies pop up in the client code (libnetwork, hashcorp/memberlist, consul, etcd ?) it means the packages in `docker/stack` are not loosely coupled. There should be no reason for the `cli` to pull those dependencies :sweat: 

A small package split / reorganization on `docker/stack` is required for this to get in `docker/cli`

Comment (alexmavr) [reply vendor.conf]: Agreed, the `stacks/pkg/interfaces` package is meant for use only by components that are internal to the engine. We'll try to unwrangle why it's being referenced here

**Comment by silvin-lubecki** [comment]
I dont' understand why all the `github.com/docker/stacks` code is not in the `github.com/moby/moby` repository, as we move the logic from the client to the server side?

**Comment by dperny** [comment]
same reason swarmkit is not in the `moby/moby` repository. more modularity.

**Comment by silvin-lubecki** [comment]
@dperny I don't see why you can't have the same modularity with a package inside moby/moby? 🤔 
A package is a package, just moving it to its own repo won't add more modularity.

File: cli/command/stack/properties.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/docker/cli/opts"
+	"github.com/docker/stacks/pkg/types"
+)
+
+func substituteProperties(input *types.ComposeInput, workingDir string) error {
```
Comment (vdemeester) [inline cli/command/stack/properties.go]: Why is this here ? Why not reusing code from `cli/cli/compose` package ? This adds yet another implementation of this variable substitution…

**Comment by thaJeztah** [comment]
> the upstream repository has a really weird line in the README.md, that is really not welcoming for contributors

@vdemeester I opened https://github.com/docker/stacks/pull/51 to make that messaging a bit more friendly (YOLO) 😇😄

**Comment by dhiltgen** [comment]
Lets close this until we sort out what the path forward is with https://github.com/moby/moby/issues/38872
